driver-core/src/main/java/com/datastax/driver/core/FramingFormatHandler.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.driver.core;
+
+import com.datastax.driver.core.Message.Response.Type;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPipeline;
+import io.netty.handler.codec.MessageToMessageDecoder;
+import java.util.List;
+
+/**
+ * A handler to deal with different protocol framing formats.
+ *
+ * <p>This handler detects when a handshake is successful; then, if necessary, adapts the pipeline
+ * to the modern framing format introduced in protocol v5.
+ */
+public class FramingFormatHandler extends MessageToMessageDecoder<Frame> {
+
+  private final Connection.Factory factory;
+
+  FramingFormatHandler(Connection.Factory factory) {
+    this.factory = factory;
+  }
+
+  @Override
+  protected void decode(ChannelHandlerContext ctx, Frame frame, List<Object> out) throws Exception {
+    boolean handshakeSuccessful =
+        frame.header.opcode == Type.READY.opcode || frame.header.opcode == Type.AUTHENTICATE.opcode;
```

Comment:
The protocol allows `OPTIONS` as the first-ever message, so shouldn't we test for `SUPPORTED` here as well? Or maybe v5 does not switch to modern framing for that first `SUPPORTED` response? (I must confess I don't remember, it's been a while)

Alternatively, we could test for `opcode != ERROR` like in the original code.

Reply:
The specs say that only READY and AUTHENTICATE triggers trigger a framing format change:

> The initial STARTUP message and any OPTIONS messages which precede it are expected to be unframed. Likewise, the responses returned by the server, SUPPORTED in response to OPTIONS and either READY or AUTHENTICATE in response to STARTUP are transmitted unframed. After sending the READY or AUTHENTICATE response to a STARTUP message, the server will begin encoding and decoding all further transmissions according to the protocol version of that STARTUP message.

That's why I figured that detecting those 2 response types would make more sense than detecting an ERROR response.

Reply:
Right, we're good then 👍